src/main/java/org/testng/internal/MethodHelper.java

```diff
@@ -237,8 +241,19 @@ public int compare(Node<ITestNGMethod> o1, Node<ITestNGMethod> o2) {
       String[] methodsDependedUpon = m.getMethodsDependedUpon();
       String[] groupsDependedUpon = m.getGroupsDependedUpon();
       if (methodsDependedUpon.length > 0) {
-        ITestNGMethod[] methodsNamed =
-          MethodHelper.findDependedUponMethods(m, methods);
+        ITestNGMethod[] methodsNamed = null;
+        // Method has instance
+        if (m.getInstance() != null) {
+          // Get other methods with the same instance
+          ArrayList<ITestNGMethod> instanceMethods = testInstances.get(m.getInstance());
+          ITestNGMethod[] allInstanceMethods = 
+              instanceMethods.toArray(new ITestNGMethod[instanceMethods.size()]);
+          // Search for other methods that depends upon with the same instance
+          methodsNamed = MethodHelper.findDependedUponMethods(m, allInstanceMethods);

```

Comment:
Maybe you can create a new findDependedUponMethods method with List instead of array param.